crates/starlight/src/vm/context.rs

```diff
@@ -188,8 +186,8 @@ impl GcPointer<Context> {
         def_native_property!(self, global_object, name.intern(), constructor)?;
 
         unsafe {
-            VM_NATIVE_REFERENCES.push(T::class() as *const _ as _);
-            VM_NATIVE_REFERENCES.push(T::raw_constructor as _);
+            self.vm.external_references.push(T::class() as *const _ as _);
```

Comment:
vm.external_references should be &'static [usize] so no one can modify it and it will be alive for entire VM lifetime.

Reply:
If someone wants to add a class to make it serializable it should pass pointer to class and ctor when VM is created. Otherwise there should be no way to extend native references array.

Reply:
Here mainly to support register custom class. And support custom class serialize and deserialize in snapshot.

Reply:
like this
```
ctx.register_external_reference(Person::say_hello as _);
let buf = Snapshot::take_context(false, &mut runtime, ctx, |_, _| {}).buffer;
let mut ctx = Deserializer::deserialize_context(&mut runtime, false, &buf);
```
We could also support snapshot custom classes

Reply:
When you create VirtualMachine you pass reference to say_hello or Person class to external_references array. register_external_reference is bad because it can easily break order of references which might cause segfault.

Reply:
Here <https://github.com/Starlight-JS/starlight/blob/dev/crates/starlight/src/vm.rs#L215>. All the external references that are not part of VM should be passed through this array and no other way to register a reference should exist.

Reply:
yeah, that makes sense. But we may need to register custom classes dynamically, then we must pass all possible references to external_references when creating virtualmachine. Also, users must pass `T::class()` and `T::raw_constructor` which are not exposed to users into external_references. Last, if we only `register_external_references`, we will not change the order of native references, and will not cause segment fault occasionally.

Reply:
Thing is list of references must be known to the VM ahead of time. No dynamic register of them. That's what V8 does because it is really unsafe to register references dynamically. And if you register reference you should not then unregister it.

Reply:
What do you think of deleting `remove_reference` and reserve `register_external_reference`? then, if a reference was registered, it will not be unregistered.

Reply:
that might work but I still don't think having register_external_reference is a good idea... I would mark it as deprecated so users would not use it.

Reply:
ok, I will remove `register_externa_reference` if there is better way to register custom class later.